### ISSUE TITLE
Finalize loading local data from .jsonl files and facilitating data blends

### ISSUE THREAD

**Issue by J38** [stanford-crfm/mistral#111]
This is currently implemented in `pubmed` branch. Feel free to comment on best way to achieve this!

**Comment by J38** [comment]
@dlwh So you wanted to change how the local data is loaded as well? It sounded like you think we should revert to the original logic for dataset construction and handle all of this in a custom data loading script? So it makes sense to merge the two objectives of 1.) increase flexibility to data loading to allow for local 2.) allow for data blends with specified mixing probs ?

**Comment by dlwh** [comment]
Yeah that's what I was thinking

**Comment by J38** [comment]
I've thought about this a bit, reviewed HF dataset loading a bit, and am guessing at what you had in mind, ... some thoughts:

So high level goals:

* support super easy dataset loading from by just specify directory and label ... (e.g. "/u/scr/nlp/data/my-math-dataset", "my-math-dataset")
* use HF dataset loading script standard
* support linking metadata to datasets
* support users custom HF data loading scripts

So I guess we want a `mistral_loading_script.py` that integrates the dataset loading logic ? And we should have that logic also search for some metadata in the directory to construct the model card? If user doesn't specify metadata some reasonable generic thing is put in place.

Another approach is that `mistral_loading_script.py` could be more like a template, and we modify it to auto-generate an experiment specific loading script that can be a record of the dataset construction for this experiment.

I am not entirely sure if the data blending belongs in `mistral_loading_script.py` or not ...

**Comment by J38** [comment]
Good resource: https://huggingface.co/docs/datasets/add_dataset.html

**Comment by J38** [comment]
Or I guess we do something like

```
mistral create-dataset --name my-math --directory /u/scr/nlp/data/my-math
```

That would auto-generate the data loading script and put in a canonical place.

Then when you run a language model training process you'd just specify:

```
name: wikitext,my-math
```

**Comment by J38** [comment]
What I'm taking away is that if a user has a directory with a bunch of `.jsonl` files and a name, we want them to create a "formal" HF dataset. So if I have some math text in `.jsonl` I should be able to hit enter and get an HF loading script like "my_math.py", and `corpora/auto.py` should just assume everything is a formal HF dataset and generically load things. And the load script acts as a record of the dataset.

**Comment by J38** [comment]
What I'm a bit unsure of is how to handle the `name="my_math,my_physics"` , `proportions=[0.5,0.5]` case. Are we aiming to create a `my_math=0.5+my_physics=0.5.py` and this somehow will be a new dataset? I follow what's going on in your example data loading script, but not sure about doing more sophisticated things like this in a standard data loading script. Do you have an example in mind?

**Comment by J38** [comment]
Will look into this, not sure if there is a way to tell HF datasets to look at certain path for local data loading scripts to facilitate `datasets.load_dataset("my_math", data_dir="...")` and it will look at `/mistral/data_scripts/my_math.py` or if we have to add that path in `auto.py` ...

**Comment by J38** [comment]
TLDR
1. Do you mean `mistral_load.py` or `my_math.py` ?
2. Does data blending happen in `corpora/auto.py` or `mistral_load.py` or `my_math=0.5+my_physics=0.5.py` ?

**Comment by dlwh** [comment]
To me, the answer to (2) is definitely `corpora/auto.py`. 

I agree that HF is terrible for ad-hoc weighting, so I think we should support static datasets and allow ad-hoc mixing in mistral itself. That is, files->sequences is done via HF datasets, but the interleaving is done by us, akin to what your branch already does.

I basically just want to replace the body of your for loop with something that only calls load_dataset, with appropriate massaging. I think it would be a good idea to add the path variable, as you suggest. (I guess I also wanted to not use split on comma but that's just because it's aesthetically displeasing to me... :-) 

For (1) I think making a mistral_load isn't a bad idea for quick ad-hoc things, but I think if we have reproducibility in mind we should ideally strive to use HF datasets and actually push them to the HF dataset repo.

**Comment by J38** [comment]
Great I'll make the changes you suggest and put up a pull request ...

**Comment by J38** [comment]
I think there is still some debate about best design choices for the interface ...

I think we want the user to be able to specify in the config:

```
id: wikitext,my_local_math
data_dir: ,/u/scr/nlp/data/my_math
ratios: 0.5,0.5
```

And `my-local-math` will get turned into `{loading_scripts_path}/my_local_math.py`

The user can manually write `my_local_math.py` if they want things configured a certain way, or we can provide a tool that auto-generates that loading script with generic content but expects the directory/files to fit a certain format.

So I'm thinking something like

```
mistral create-dataset --name my_local_math --path /u/scr/nlp/data/my_local_math
```

Or it could look like:

```
id: wikitext,mistral
data_dir: ,/u/scr/nlp/my_math_data
ratios: 0.5,0.5
```

And we create a custom loading script `mistral_data_load.py` that assumes a canonical directory structure/file format ... e.g. "train" in training files, "validation" in validation files.

And it sounds like Tony wants to give the user more choice in how their directory is set up. And I'm not sure if we'd want to allow a list of data_files (another arg we can pass to load_datasets) to play a role as well.

I will think this over some more ...

**Comment by J38** [comment]
I guess we could support both workflows:

1.) For fast ad hoc experimentation, just use the standardized mistral loading script and throw your files into a reasonable directory structure/naming convention and it will just work ... maybe we should let users specify data_files for each dataset so they can also quickly specify how they want their directory set up

2.) CLI that will create a generic data loading script that the user can customize to encourage more formality

**Comment by J38** [comment]
So looked over #112 some more and thinking about this more ...

1.) We should stick to the Hugging Face way of doing things as much as possible

2.) We can make a CL tool that'll make it super easy to create a standard HF datasets loading script ... basically automate what David did for pubmed, and it'll place the script in a reasonable place like `scripts` dir ... so something like:

```
mistral create-dataset --name my_math --data_files {"train": ["train.txt"], "validation": ["validation.txt"]}
```

3.) If you create a new dataset, which you are encouraged to do, you can just refer to it by its name and Mistral will look in the canonical (or you can configure) data loading scripts dir and it work as if you put "wikitext"

4.) Alternatively, we can do something kind of like in #112 and you can just directly specify lists in the *.yaml for your experiment that will use the HF load_dataset args data_dir and _data_files:

So maybe it would look something like this in the config (I'll figure out how to make this work):

```
path:
    - wikitext
    - jsonl
    - jsonl

name:
    - wikitext-103
    - my_math
    - my_physics

data_dir:
    -
    - /u/scr/nlp/data/math
    - /u/scr/nlp/data/physics

data_files:
    train:
        -
        - my_math_train.jsonl,other_math_train.jsonl
        - my_physics_train.jsonl,other_physics_train.jsonl
    validation:
        -
        - my_math_validation.jsonl,other_math_validation.jsonl
        - my_physics_validation.jsonl,other_physics_validation.jsonl

data_ratios:
    - 0.20
    - 0.40
    - 0.40
```

**Comment by J38** [comment]
And maybe next level would be:

```
data_files:
    train:
        -
        - *train*
        - *train*
     validation:
         -
         - *validation*
         - *validation*
```

**Comment by dlwh** [comment]
I'm definitely pro (1) (2) and (3). I kind of instinctively dislike (4) as structured, but really only because I want to transpose it to something more like:

```yaml
datasets:
  wikitext:
      train: *paths*
      validation: **
      weight: 0.3
   my_math:
      # etc
```

or maybe it's paths to included yaml files or something. Basically I'm just arguing that all the stuff for wikitext should be in one place, rather than scattered about where it's easy to miss.

**Comment by J38** [comment]
Oh okay that's a good point, so it should be organized per dataset I agree that's better. I'll put something together and see what you think.

**Comment by J38** [comment]
Okay I implemented this and am checking for correctness:

```
# blended_demo.yaml
#   An even blend of wikitext-2 + pile_pubmed
---
dataset:

    id: blend-demo-1

    datasets:
        - path: wikitext
          name: wikitext-2-raw-v1
        - path: pile_pubmed
          name: all
          data_dir: /nlp/scr/jebolton/pile_pubmed

    dataset_ratios: [0.50, 0.50]

    validation_ratio: null

    # Number of Preprocessing Workers
    num_proc: 4

    # Number of Evaluation Preprocessing Workers
    eval_num_proc: 4
```

**Comment by J38** [comment]
Note that I added an id for the combination so you can cache the combo dataset.

**Comment by J38** [comment]
Also you can do:

```
datasets:
    - path: my-math
       train: ["train.txt"]
       validation: ["validation.txt"]
       data_dir: /u/scr/nlp/data/my_math
```

... it will load train and validation into a dictionary for the `data_files` argument ...

alternatively we could do:

```
datasets:
    - path: my-math
      data_files:
          train: ["train.txt"]
          validation: ["validation.txt"]
      data_dir: /u/scr/nlp/data/pile_pubmed
```

which might make it clearer train and validation are going to be passed to `data_files` ... not sure which I like better ...

**Comment by dlwh** [comment]
looks great! 

at some point we should probably figure out a better preprocessing solution than what we have now and possibly move more towards streaming a la mosaic, but we can cross that bridge later.

**Comment by percyliang** [comment]
Might it be better to include the ratios as fields in the dataset list?  If we have 10 datasets, worried that it'll be somewhat hard to read / error prone to associate the elements of the array with the actual datasets.

**Comment by dlwh** [comment]
yeah, i went back and forth on that.  I think  we could call it "weight" rather than "ratio" and put it like you say that's probably the easiest thing to support, though maybe the semantics wouldn't be super clear.

**Comment by J38** [comment]
And in addition to this we want to support  `python train.py --data /u/scr/nlp/data/my-math` and it will reasonably interpret the directory, for instance inspect and determine file types (which I think HF handles), naturally split into train/validation (or look for "train" and "validation" in file names) and maybe auto determine number of devices and just auto be set to train a gpt2-small

**Comment by J38** [comment]
Also a note, if you put nothing for data_ratios it will just combine all of them ...

**Comment by J38** [comment]
Okay so give each entry a weight, and those will be combined into the appropriate ratios, that seems like a good idea ...

**Comment by dlwh** [comment]
another reasonable option is to make data_ratios a dict with keys equal to the ids of the datasets

**Comment by percyliang** [comment]
I like the dict idea.

**Comment by dlwh** [comment]
not doing blends anymore.
